tests/Functions/SelectServerFunctionalTest.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace MongoDB\Tests\Functions;
+
+use MongoDB\Driver\ReadPreference;
+use MongoDB\Driver\Server;
+use MongoDB\Tests\FunctionalTestCase;
+
+use function MongoDB\select_server;
+
+class SelectServerFunctionalTest extends FunctionalTestCase
+{
+    /** @dataProvider providePinnedOptions */
+    public function testSelectServerPrefersPinnedServer(array $options): void
+    {
+        $this->skipIfTransactionsAreNotSupported();
+
+        if (! $this->isShardedCluster()) {
+            $this->markTestSkipped('Pinning requires a sharded cluster');
+        }
+
+        if (! $this->isLoadBalanced()) {
+            $this->markTestSkipped('libmongoc does not pin for load-balanced topology');
+        }
+
+        /* By default, the Manager under test is created with a single-mongos
+         * URI. Explicitly create a Client with multiple mongoses. */
+        $client = static::createTestClient(static::getUri(true));
+
+        // Collection must be created before the transaction starts
+        $this->createCollection($this->getDatabaseName(), $this->getCollectionName());
+
+        $session = $client->startSession();
+        $session->startTransaction();
+
+        $collection = $client->selectCollection($this->getDatabaseName(), $this->getCollectionName());
+        $collection->find([], ['session' => $session]);
+
+        $this->assertTrue($session->isInTransaction());
+        $this->assertInstanceOf(Server::class, $session->getServer(), 'Session is pinned');
+        $this->assertEquals($session->getServer(), select_server($client->getManager(), ['session' => $session]));
```

Comment:
Note: this test is still a bit dubious, as we don't have any insight into what `Manager::selectServer()` is _actually_ doing. That may change with server selection logging ([PHPLIB-1000](https://jira.mongodb.org/browse/PHPLIB-1000)).

Currently, there's a chance that server selection just happens to select the same server that was pinned (see: [Topology type: Sharded](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-sharded) in the Server Selection spec). Supplying an alternative read preference in the operation-level options doesn't really affect this, since it wouldn't influence mongos selection. Again, logging would help verify that.

Having said all that, I think adding a functional test is a good first step since we previously had no direct test coverage for this function. This file at least provides a starting point to add more coverage down the line. I've left a comment in [PHPLIB-1000](https://jira.mongodb.org/browse/PHPLIB-1000) to that effect.